node/src/coordinator.rs

```diff
@@ -317,29 +342,155 @@ impl Coordinator {
+                        EcdsaTaskId::KeyGeneration {
+                            key_event: contract_event.id.clone(),
+                        },
+                        network_client.all_participant_ids(),
+                    )?;
+                    tracing::info!(
+                        "opened channel, sending start keygen on id: {:?}",
+                        channel.task_id()
+                    );
+                    chain_txn_sender
+                        .send(ChainSendTransactionRequest::StartKeygen(
+                            ChainStartKeygenArgs {},
+                        ))
+                        .await?;
+                    tracing::info!("sent start keygen, starting computation");
+                    // note: we could inline below function, but maybe better to wait and see how other
+                    // singauter schemes will be handled.
+                    let res = EcdsaSignatureProvider::run_key_generation_client(
+                        mpc_config.participants.threshold as usize,
+                        channel,
+                    )
+                    .await;
+                    let res = match res {
+                        Ok(res) => res,
+                        Err(e) => anyhow::bail!("error: {}", e),
+                    };
+                    tracing::info!("leader concluded computation, storing keyshare.");
+                    let keyshare = Keyshare {
+                        key_id: contract_event.id.clone(),
+                        data: KeyshareData::Secp256k1(res.clone()),
+                    };
+                    keyshare_storage.store_key(keyshare).await?;
+                    tracing::info!("leader stored keyshare.");
+                    let my_public_key = affine_point_to_public_key(res.public_key)?;
+                    let key_event_id = contract_event.id;
+                    while contract_event.completed.len() != n_participants - 1 {
+                        tracing::info!("leader waiting on follower votes");
+                        contract_event = key_event_receiver.borrow_and_update().clone();
+                        tokio::time::sleep(std::time::Duration::from_millis(500)).await;
+                    }
+                    tracing::info!("leader got follower votes");
+                    if contract_event.id != key_event_id {
+                        tracing::info!("Key generation timed out.");
+                        continue 'leader;
+                    }
+                    tracing::info!("Key generation complete; Leader calls vote_pk.");
+                    chain_txn_sender
+                        .send(ChainSendTransactionRequest::VotePk(ChainVotePkArgs {
+                            key_event_id: contract_event.id,
+                            public_key: my_public_key,
+                        }))
+                        .await?;
+                } else {
+                    // note: if leader does not exit here, we just hang. Not sure why.
```

Comment:
I think if we're leader and we see that there is already an event started, we should send vote_abort (lol oops that doesn't even exist on the contract!) and then just sleep and loop. Eventually it will be aborted or time out and then we can start the next attempt.

Reply:
#297